### PR TITLE
Add `max_by` and `min_by` aggregation functions

### PR DESCRIPTION
**Pull request by mfussenegger**
## Summary of the changes / Why this improves CrateDB

Closes https://github.com/crate/crate/issues/10528

## Checklist

 - [x] Added an entry in `CHANGES.txt` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in ``CHANGES.txt``
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

File: docs/general/builtins/aggregation.rst [hunk 1]
```diff
@@ -519,6 +519,45 @@ Some Examples::
...
+An Example::
+
+    cr> SELECT min_by(mountain, height) FROM sys.summits;
+    +--------------------------+
+    | min_by(mountain, height) |
+    +--------------------------+
+    | Puy de Rent              |
+    +--------------------------+
+    SELECT 1 row in set (... sec)
+
```
Comment (matriv) [inline docs/general/builtins/aggregation.rst:559]: Maybe some NOTE about nulls, and ties?

File: server/src/test/java/io/crate/execution/engine/aggregation/impl/CmpByAggregationTest.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            DataTypes.UNTYPED_OBJECT.getTypeSignature(),
+            DataTypes.STRING.getTypeSignature()
+        );
+        assertThatThrownBy(() -> executeAggregation(
+            signature,
+            new Object[][] {
+                new Object[] { "foo", Map.of("x", 10) },
+            },
+            List.of()
+        )).hasMessage("Cannot use `max_by` on values of type object");
```
Comment (matriv) [inline server/src/test/java/io/crate/execution/engine/aggregation/impl/CmpByAggregationTest.java:93]: minor: maybe also check exception class.

File: server/src/test/java/io/crate/execution/engine/aggregation/impl/CmpByAggregationTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+import org.junit.Test;
+
+import io.crate.metadata.functions.Signature;
+import io.crate.operation.aggregation.AggregationTestCase;
+import io.crate.types.DataTypes;
+
+public class CmpByAggregationTest extends AggregationTestCase {
+
+    @Test
```
Comment (matriv) [inline server/src/test/java/io/crate/execution/engine/aggregation/impl/CmpByAggregationTest.java:38]: Can we have some more tests?
- If all values of `searchField` are null
- If all values of `returnField` are null (shouldn't be an edge case though)
- If both?
- if we have tied values on `searchField` and different values for those ties in `returnField`?

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+import io.crate.types.DataTypes;
+import io.crate.types.TypeSignature;
+
+public final class CmpByAggregation extends AggregationFunction<CmpByAggregation.CompareBy, Object> {
+
+    public static final String MAX_BY = "max_by";
+    public static final String MIN_BY = "min_by";
+
+    static class CompareBy {
+
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java:50]: :P

**Review by matriv — COMMENTED** [review]
Thank you! Left a few comments around edge cases with ties and nulls.

**Review by matriv — APPROVED** [review]
LGTM. Thank you!
One last comment, if you think is not too much:
For those null/tie cases, can we have a for loop and test both signatures?

**Comment by mfussenegger** [comment]
> For those null/tie cases, can we have a for loop and test both signatures?

Sure we could, but what for? The only difference between min and max is `cmp -> cmp == 1` vs `cmp -> cmp == -1` and that's already covered

**Comment by matriv** [comment]
> > For those null/tie cases, can we have a for loop and test both signatures?
> 
> Sure we could, but what for? The only difference between min and max is `cmp -> cmp == 1` vs `cmp -> cmp == -1` and that's already covered

For sure, only for "completeness", but functionally doesn't provide any value.
Maybe you could do a `random.from("max_by", "min_by")` and make the test method names more generic, but as you like, not insisting at all, more like OCD driven :)

**Comment by mfussenegger** [comment]
Added another fixup - your comment gave me the idea to avoid the cmp function call and I also extended the tests while I was at it.

File: server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java [hunk 5]
```diff
@@ -71,7 +71,7 @@ public static void register(AggregationImplModule mod) {
                 variableConstraintA,
                 variableConstraintB
             ),
-            (signature, boundSignature) -> new CmpByAggregation(cmp -> cmp == 1, signature, boundSignature)
+            (signature, boundSignature) -> new CmpByAggregation(1, signature, boundSignature)
```
Comment (matriv) [inline server/src/main/java/io/crate/execution/engine/aggregation/impl/CmpByAggregation.java:74]: nice!

**Review by matriv — APPROVED** [review]
Thx again!

**Pull request merged**